### PR TITLE
Implemented GIoU

### PR DESCRIPTION
**Pull request by briankosw**
## What does this PR do?

Implements Generalized Intersection over Union as mentioned in #251

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by akihironitta** [comment]
@briankosw Thank you for your contribution! Let us know if you need help :]

**Comment by briankosw** [comment]
Hi @akihironitta! I'm trying to write tests for GIoU, but I don't have experience with `unittest` (I'm more used to `pytest`) and it seems that the majority of the tests in this repo weren't fully setup. I'll submit a commit soon, and could you take a look at review it for me?

**Comment by akihironitta** [comment]
@briankosw Sure! Mark this PR as ready for review when you're ready :]

File: pl_bolts/losses/giou.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    Calculates the generalized intersection over union loss.
+
+    Args:
+        pred: batch of prediction bounding boxes with representation
+              [x_min, y_min, x_max, y_max]
+        target: batch of target bounding boxes with representation
+                [x_min, y_min, x_max, y_max]
+
+    Returns:
+        loss
```
Comment (ydcjeff) [inline pl_bolts/losses/giou.py:19]: rather be `GIoU` loss instead of `loss`?
```suggestion
        GIoU loss
```

File: tests/losses/test_giou.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+from unittest import TestCase
+import torch
+
+from pl_bolts.losses.giou import giou_loss
+
+
+class TestGIoULoss(TestCase):
+    def test_complete_overlap(self):
+        pred = torch.tensor([[100, 100, 200, 200]])
```
Comment (ydcjeff) [inline tests/losses/test_giou.py:13]: can we test with 10, 20 instead of 100, 200?

**Review by ydcjeff — COMMENTED** [review]
Great to see GIoU loss in bolts!
Can you add the docs?

File: tests/losses/test_giou.py (hunk 2 above)
Comment (briankosw) [reply tests/losses/test_giou.py:13]: So I actually separated the test cases into three separate categories:
- cases where the prediction bbox and target bbox are completely overlapping (i.e. perfect prediction)
- cases where the prediction bbox and target bbox are completely non-overlapping (i.e. completely missed prediction)
- cases where the prediction bbox and target bbox are slightly overlapping

The test case you're suggesting would be slightly overlapping as opposed to completely overlapping. Do you think I should add the new test case you're suggesting or what do you think?

Comment (ydcjeff) [reply tests/losses/test_giou.py:13]: I see! ok current tests are good

File: pl_bolts/losses/giou.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                [x_min, y_min, x_max, y_max]
+
+    Returns:
+        GIoU loss
+    """
+    eps = 1e-6
+    x_min = torch.max(pred[:, 0], target[:, 0])
+    y_min = torch.max(pred[:, 1], target[:, 1])
+    x_max = torch.min(pred[:, 2], target[:, 2])
+    y_max = torch.min(pred[:, 3], target[:, 3])
```
Comment (akihironitta) [inline pl_bolts/losses/giou.py:25]: `min` and `max` seem swapped?

**Review by akihironitta — CHANGES_REQUESTED** [review]
Thank you for the swift action! Would you mind having a look?

File: pl_bolts/losses/giou.py (hunk 3 above)
Comment (briankosw) [reply pl_bolts/losses/giou.py:25]: This is actually intended behavior! `x_min` refers to the minimum x value for the intersection between the prediction bbox and the target bbox, and you derive that value by getting the max between the `x_min` of the prediction bbox and the `x_min` of the target bbox.

Comment (briankosw) [reply pl_bolts/losses/giou.py:25]: I'd refer you to [Generalized Intersection over Union](https://giou.stanford.edu/), which contains super useful visualizations and explanations for how it works 🙂

Comment (akihironitta) [reply pl_bolts/losses/giou.py:25]: I see, then let's keep it as is!

**Review by akihironitta — APPROVED** [review]
LGTM!

**Comment by briankosw** [comment]
I'm going to add a little bit more to the docstring and fix that isort failure, but that should be the final touches!

File: pl_bolts/losses/giou.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    Args:
+        pred: batch of prediction bounding boxes with representation
+              [x_min, y_min, x_max, y_max]
+        target: batch of target bounding boxes with representation
+                [x_min, y_min, x_max, y_max]
+
+    Returns:
+        GIoU loss
+    """
+    eps = 1e-6
```
Comment (Borda) [inline pl_bolts/losses/giou.py:24]: shall this go to constants? this would be depending on if we use AMP (float16) right?

File: tests/losses/test_giou.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+"""
+
+from unittest import TestCase
+
+import torch
+
+from pl_bolts.losses.giou import giou_loss
+
+
+class TestGIoULoss(TestCase):
```
Comment (Borda) [inline tests/losses/test_giou.py:12]: can we rather write it as simple `test_complete_overlap` with parameterized inputs and expected results?
it would be easier to debug later as we would si what exact configurations fails

File: pl_bolts/losses/giou.py (hunk 4 above)
Comment (briankosw) [reply pl_bolts/losses/giou.py:24]: Ah good point. Should I extract this out as a global constant or as a passed-in argument?

File: tests/losses/test_giou.py (hunk 5 above)
Comment (briankosw) [reply tests/losses/test_giou.py:12]: I'm unfamiliar with `unittest` (I'm used to `pytest`) so I didn't realize you can do that in `unittest` as well!

Comment (briankosw) [reply tests/losses/test_giou.py]: I realized that `pytest` is being used in conjunction with `unittest`, so I just used `pytest` to parameterize the tests.

**Comment by akihironitta** [comment]
Also, please remember to add the summary of your change to the changelog.

File: tests/losses/test_giou.py (hunk 5 above)
Comment (Borda) [reply tests/losses/test_giou.py]: happy to help if you want, just ping me on slack :]

File: pl_bolts/losses/giou.py (hunk 4 above)
Comment (Borda) [reply pl_bolts/losses/giou.py:24]: well in fact the question is if we really need to use eps as the eps is needed only for zeros decision...

**Review by Borda — COMMENTED** [review]
pls follow @akihironitta suggestions

**Comment by annikabrundyn** [comment]
This is awesome! Does it make sense to also add this to the PL metrics?

**Comment by Borda** [comment]
> This is awesome! Does it make sense to also add this to the PL metrics?

I would keep it here for a little while and later like any other metric we may move them to PL, we just need to try they are stable and kind of standard :]

**Comment by ananyahjha93** [comment]
@annikabrundyn @briankosw @Borda let's add this to PL as a class metric + the current functional form.

**Comment by briankosw** [comment]
I'd be happy to close this PR and open a new one in PL as soon as I'm given the go ahead 🙂

**Comment by Borda** [comment]
> I'd be happy to close this PR and open a new one in PL as soon as I'm given the go ahead 🙂. If so, I'm assuming #370 and part of #251 should be migrated over as well?

hi, the point is that we shall merge it here, and later (let's say a few weeks) we add it in PL and eventually drop from here

**Comment by briankosw** [comment]
I think this PR is ready for a final review. Would love your reviews @Borda @ydcjeff.

**Comment by ydcjeff** [comment]
Thank you for your work @briankosw 
Is there any difference between this and [this one from torchvision](https://pytorch.org/docs/stable/torchvision/ops.html#torchvision.ops.generalized_box_iou) except that we extract giou from 1 in this?

**Comment by briankosw** [comment]
> Thank you for your work @briankosw
> Is there any difference between this and [this one from torchvision](https://pytorch.org/docs/stable/torchvision/ops.html#torchvision.ops.generalized_box_iou) except that we extract giou from 1 in this?

Ah I didn't realize torchvision has its own implementation. You're correct in that we subtract giou from 1. Should I change it to use torchvision's giou?

**Comment by ydcjeff** [comment]
> > Thank you for your work @briankosw
> > Is there any difference between this and [this one from torchvision](https://pytorch.org/docs/stable/torchvision/ops.html#torchvision.ops.generalized_box_iou) except that we extract giou from 1 in this?
> 
> Ah I didn't realize torchvision has its own implementation. You're correct in that we subtract giou from 1. Should I change it to use torchvision's giou?

I think both are right. in giou loss, we subtract giou from 1. Since torchvision has giou, we only need to 1 - giou to get giou loss... So it can be said this PR is unneeded, Thank you for your work btw.

But let's see core members' thoughts too!

**Comment by akihironitta** [comment]
@briankosw @ydcjeff
Thank you for your work! In my opinion, it's still nice to have the loss here or in PL as the current function and also a subclass of `pl.metrics.Metric` like `pl.metrics.Accuracy` as @ananyahjha93 suggested.

> @annikabrundyn @briankosw @Borda let's add this to PL as a class metric + the current functional form.

I think it's reasonable to have the loss in Bolts and later migrate it to PL like @Borda mentioned.

@Borda @ananyahjha93 So, how shall we move on?

**Comment by Borda** [comment]
yes, as we do not have any loss in PL, we shall also think if it is reasonable to include it there... eventually we would move the losses as a sub-package... so lets it land here and later talk about losses as a complete functional block :]

File: pl_bolts/losses/object_detection.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    Calculates the generalized intersection over union loss.
+
+    It has been proposed in `Generalized Intersection over Union: A Metric and A
+    Loss for Bounding Box Regression <https://arxiv.org/abs/1902.09630>`_.
+
+    Args:
+        pred: batch of prediction bounding boxes with representation
+              [x_min, y_min, x_max, y_max]
+        target: batch of target bounding boxes with representation
+                [x_min, y_min, x_max, y_max]
```
Comment (Borda) [inline pl_bolts/losses/object_detection.py:19]:
```suggestion
        pred: batch of prediction bounding boxes with representation ``[x_min, y_min, x_max, y_max]``
        target: batch of target bounding boxes with representation ``[x_min, y_min, x_max, y_max]``
```
minor formating

File: pl_bolts/losses/object_detection.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    Returns:
+        GIoU loss
+    """
```
Comment (Borda) [inline pl_bolts/losses/object_detection.py:23]: pls add doctest example

File: pl_bolts/losses/object_detection.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+"""
+Generalized Intersection over Union (GIoU) loss (Rezatofighi et. al)
+"""
+
+import torch
+
+
+def giou_loss(pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
```
Comment (Borda) [inline pl_bolts/losses/object_detection.py:8]: can we define separately `giou` as functional metric and here use as "supplement"?
so we gain more from this addition :]

**Review by Borda — COMMENTED** [review]
lets add giou as metric and reuse it in loss

File: pl_bolts/losses/object_detection.py (hunk 7 above)
Comment (briankosw) [reply pl_bolts/losses/object_detection.py:22]: Forgive my ignorance, but how do I do that?

Comment (akihironitta) [reply pl_bolts/losses/object_detection.py:22]: I think @Borda means something like:
https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/58536c260e7877c8c75dfccfaa58acee4fdc2ae1/pl_bolts/datasets/cifar10_dataset.py#L36-L50

**Comment by briankosw** [comment]
It seems that there is no documentation for metrics in pl_bolts. Should I go ahead and create one? @akihironitta

**Review by Borda — APPROVED** [review]
